includes/functions.php

```diff
@@ -1593,3 +1593,143 @@ function is_self_ping( $id ) {
 
 	return false;
 }
+
+/**
+ * Get the embed of a URL.
+ *
+ * This function supports both ActivityPub JSON and oEmbed URLs.
+ * It first tries to fetch ActivityPub JSON, and if that fails,
+ * falls back to oEmbed.
+ *
+ * @param string $url The URL to get the embed of.
+ *
+ * @return string|false The embed of the URL or false if not found.
+ */
+function embed_get( $url ) {
+	$embed = \wp_oembed_get( $url );
+	if ( $embed ) {
+		return $embed;
+	}
+
+	// Create unique transient keys for both HTTP calls.
+	$main_transient_key = '_activitypub_embed_' . md5( $url );
+	$embed              = \get_transient( $main_transient_key );
+
+	if ( false === $embed ) {
+		$response = Http::get( $url );
+		if ( \wp_remote_retrieve_response_code( $response ) === 200 ) {
+			$embed = \json_decode( \wp_remote_retrieve_body( $response ), true );
+			\set_transient( $main_transient_key, $embed, DAY_IN_SECONDS );
+		} else {
+			return \wp_oembed_get( $url );
+		}
+	}
+
+	if ( $embed ) {
+		$author_name = isset( $embed['attributedTo'] ) ? $embed['attributedTo'] : '';
+		$author_url  = $author_name;
+		$avatar_url  = isset( $embed['icon']['url'] ) ? $embed['icon']['url'] : '';
+
+		// If we don't have an avatar URL but we have an author URL, try to fetch it.
+		if ( ! $avatar_url && $author_url ) {
+			$author_transient_key = '_activitypub_author_' . md5( $author_url );
+			$author               = \get_transient( $author_transient_key );
+
+			if ( false === $author ) {
+				$author_response = Http::get( $author_url );
+				if ( \wp_remote_retrieve_response_code( $author_response ) === 200 ) {
+					$author = \json_decode( \wp_remote_retrieve_body( $author_response ), true );
+					\set_transient( $author_transient_key, $author, DAY_IN_SECONDS );
+				}
+			}
+
+			if ( $author ) {
+				if ( isset( $author['icon']['url'] ) ) {
+					$avatar_url = $author['icon']['url'];
+				}
+				if ( isset( $author['name'] ) ) {
+					$author_name = $author['name'];
+				}
+			}
+		}
+
+		$published = isset( $embed['published'] ) ? \gmdate( get_option( 'date_format' ) . ', ' . get_option( 'time_format' ), \strtotime( $embed['published'] ) ) : '';
+		$title     = isset( $embed['name'] ) ? $embed['name'] : '';
+		$content   = isset( $embed['content'] ) ? $embed['content'] : '';
+		$boosts    = isset( $embed['shares']['totalItems'] ) ? (int) $embed['shares']['totalItems'] : 0;
+		$favorites = isset( $embed['likes']['totalItems'] ) ? (int) $embed['likes']['totalItems'] : 0;
+
+		$image = '';
+		if ( isset( $embed['image']['url'] ) ) {
+			$image = $embed['image']['url'];
+		} elseif ( isset( $embed['attachment'] ) ) {
+			foreach ( $embed['attachment'] as $attachment ) {
+				if ( isset( $attachment['type'] ) && 'Document' === $attachment['type'] ) {
+					$image = $attachment['url'];
+					break;
+				}
+			}
+		}
+
+		ob_start();
+		?>
+		<div class="activitypub-embed">
```

Comment:
why not use a template and `load_template` here? (I am not a big fan of inline HTML)